### PR TITLE
Fix syntax error on JSON range formatting

### LINKED ISSUE #4009: Errors when range does not include curly brackets (via rangeStart and rangeEnd) in json files using node api
Hello Prettier team and thank you for all the hard work you put into this awesome tool. Truly amazing work! 🥇 

I have an issue to report though. When formatting json files using the node api and providing rangeStart and rangeEnd values not including start `{`-sign or end `}` sign it errors. 

I could not recreate it with the playground (unless I missed the rangeStart/rangeEnd options there in which case I am sorry. If it's possible just tell me and I'll put up the repro there instead).

**Prettier 1.10.2**
Reproduction repro: https://github.com/erik-lundgren/prettier-json-range-bug
```js
{
    filepath: filepath,
    rangeStart: 1, // skip start {
    rangeEnd: fileContents.indexOf("}") // skip end }
}
```

**Input:**
```json
{
    "foo": "bar"
}

```

**Output:**
```jsx
SyntaxError: Unexpected token (1:6)
> 1 | "foo": "bar"
    |      ^
    at createError (C:\Projects\prettier-json-range-bug\node_modules\prettier\
parser-babylon.js:1:110)
    at Object.parse (C:\Projects\prettier-json-range-bug\node_modules\prettier
\parser-babylon.js:1:986)
    at Object.parse$4 [as parse] (C:\Projects\prettier-json-range-bug\node_mod
ules\prettier\index.js:24274:19)
    at formatWithCursor (C:\Projects\prettier-json-range-bug\node_modules\pret
tier\index.js:33650:27)
    at format (C:\Projects\prettier-json-range-bug\node_modules\prettier\index
.js:33695:10)
    at formatRange (C:\Projects\prettier-json-range-bug\node_modules\prettier\
index.js:33922:26)
    at formatWithCursor (C:\Projects\prettier-json-range-bug\node_modules\pret
tier\index.js:33654:30)
    at format (C:\Projects\prettier-json-range-bug\node_modules\prettier\index
.js:33695:10)
    at format (C:\Projects\prettier-json-range-bug\node_modules\prettier\index
.js:33945:12)
    at Object.<anonymous> (C:\Projects\prettier-json-range-bug\index.js:8:1)
    at Module._compile (module.js:643:30)
    at Object.Module._extensions..js (module.js:654:10)
    at Module.load (module.js:556:32)
    at tryModuleLoad (module.js:499:12)
    at Function.Module._load (module.js:491:3)
    at Function.Module.runMain (module.js:684:10)
error Command failed with exit code 1.
```

**Expected behavior:**
Output formatted file, exit code 0.

**Additional debug information**
Operating system: Windows 10
yarn version: 1.3.2
node version: 8.9.4

Let me know if I can provide any more details

### LINKED ISSUE #7116: JSON range formatting throws syntax error
**Prettier 1.19.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEwAEAdEAjLS0BMaAviADQgQAOMAltAM7KgCGATmxAO4AK7CTFCwBuEWgBNyONizABrODADKVWbSgBzZDDYBXOBQAWMALYAbAOqHa8BqrBwlAm7WE2AnsnAMmFdQzg2GB4ZDRMWZAAzFjMAigArBgAPACEZeUUlFhM4ABl1OCiYuJBEpKV1DTM4AEVdCHgi2IMQVTYAti9E6CkqNnUYCwkYQ2QADgAGCj6IAIsZKi8+uA7hQooAR3r4EOpBEBYGAFooODhxc6k2OC3aa5CWMIikaOaKAJNabT0WhkrquoNQovYotGAsbBDcQjZAECg6Fi0MyVADCEBM4S8UGg6xAugCABUIYJXiVhPoAJJQC6wJRgfo0ACC1KUMHc1SaJRkmkc4KCyAArBRuRo4ABRanIACcxGIQA)
```sh
--parser json
--range-end 9
--range-start 5
```

**Input:**
```jsonc
{ "b": 2 }
```

**Output:**
```jsonc
SyntaxError: Unexpected token (1:4)
> 1 | "b": 2
    |    ^
```

**Expected behavior:**
```jsonc
{ "b": 2 }
```

This is the smallest example I could make that causes the parser to fail. It seems like if the range starts with a `:` but doesn't end with a `,` or `}` after the object value, then the parser throws a syntax error.

### PR DESCRIPTION
**Pull request by fisker**
## Description

Fixes #4009 
Fixes #7116 

Previous attempt #8082

## Checklist

- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: tests/json/range/json-stringify/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+> 2 | {a:\`a\`,n:
+    |     ^
+  3 | ''},
+  4 | {a:\`a\`,n:
+  5 | ''}
+  6 | ]
+  7 |
+=====================================output=====================================
+[
+{a:\`a\`,n:
```
[the remark this answers is not in the document]
Comment (fisker) [reply tests/json/range/json-stringify/__snapshots__/jsfmt.spec.js.snap]: Ah, it's not the reason, fixed `json5` parser in https://github.com/prettier/prettier/pull/10497/commits/a51cf97343507af9738d88720e024c018abf2acd, but still can't get ancestors in `json-stringify`

**Comment by thorn0** [comment]
Is this known?

**Prettier pr-10497**
[Playground link](https://deploy-preview-10497--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEwA6ICGAbbB5KDJAAgwGUBLKAc2zgwBpSQAZK+5ZuAD0wFsADnUZpCIKBDIDMYDkhgAnAK5wGo4hs1btOnRgCOSiPCLF0ULQHIAzlVpwAikfiWSlnvyFxLai7v8aGFQAJgjwwaaKKr4aAL4xzAIAFtBwAHJKfABGcArWpgDa6proIDAAngJyzCl8HAlaGFCZOQqmGABMAIwdxACsAwC0PQDMACwY8cUapRVV7SAQAGZLFLIifo1lCpgU2HYLkHx8mNbEWeXEMmAUobCTCQC6vrEgDIsCMBTQ1sigmAoFBAAO4ABQBCF+KBwwMw5V+7yyOzAAGs4DApDI7MgoqoQHBsnBgqFgixMDQlJhqHAAGIQBQnGBfGjILBKGAQN4gJIwPjYADqSQo8Gs0lkZEhwooADdheVWWBrAiQFRrLkYKCdtQTsgljg1e8AFbWbgAIWRaIx-DgbCgcF1+rxxu4lBodCcxntSD12ANIGkeVyrON0C5AgUVBg-NuMCSyAAHAAGd7hiBq-k7ASs8NwNUKaX296GT2aiACKFYayDO1EolchRwQwUBuaqk672O95qvgUHHKPG2N2OZxen1+mCYLLR4Kx5Add6KXb7GgAYQgx0wrNzfS5SjVABVJ1Cx3jpSoAJJQO4YsARz4AQSvZAqdAdvrxOxocGfAJgyC6ibxu8n7UgAole-5dH0sSxEAA)

```sh
--parser json
--range-end 115
--range-start 108
```

**Input:**

```jsonc
{"allOn": "Single", "Line": "example",
"noSpace":true,
                           "quote": {
    'singleQuote': 'example',
                  "indented": true,
  },
  "phoneNumbers": [
    {"type": "home",
      "number": "212 555-1234"},
    {"type": "office",
      "trailing": "commas by accident"},
  ],
}
```

**Output:**

```jsonc
{"allOn": "Single", "Line": "example",
"noSpace":true,
                           "quote": {
                             "singleQuote": "example",
                             "indented": true
                           },
  "phoneNumbers": [
    {"type": "home",
      "number": "212 555-1234"},
    {"type": "office",
      "trailing": "commas by accident"},
  ],
}
```

**Comment by fisker** [comment]
I think this is expected?

**Comment by thorn0** [comment]
You're right. Prettier can't know what indentation to use. Its only clue is the original indentation.

File: src/main/range-util.js [hunk 2]
```diff
@@ -3,6 +3,24 @@
...
+    startNodeAndParents.node,
+    ...startNodeAndParents.parentNodes,
+  ];
+  const endNodeAndAncestors = [
+    endNodeAndParents.node,
+    ...endNodeAndParents.parentNodes,
+  ];
+  return startNodeAndAncestors.find(
+    (node) =>
+      jsonSourceElements.has(node.type) && endNodeAndAncestors.includes(node)
```
Comment (thorn0) [inline src/main/range-util.js:20]: Nested loops. It's better to make `endNodeAndAncestors` a set instead of an array.

**Review by thorn0 — APPROVED** [review]

**Pull request merged**